src/clarity/clarityTypes.ts

```diff
@@ -0,0 +1,391 @@
+import BN from 'bn.js';
+import { LengthPrefixedString, Address } from '../types';
+import { CLARITY_INT_SIZE } from '../constants';
+import { BufferReader, BufferArray } from '../utils';
+
+/**
+ * Type IDs corresponding to each of the Clarity value types as described here:
+ * {@link https://github.com/blockstack/blockstack-core/blob/sip/sip-005/sip/sip-005-blocks-and-transactions.md#clarity-value-representation}
+ */
+enum ClarityType {
+  Int = '00',
+  UInt = '01',
+  Buffer = '02',
+  BoolTrue = '03',
+  BoolFalse = '04',
+  PrincipalStandard = '05',
+  PrincipalContract = '06',
+  ResponseOk = '07',
+  ResponseErr = '08',
+  OptionalNone = '09',
+  OptionalSome = '10',
+  List = '11',
+  Tuple = '12',
+}
+
+const prefixTypeID = (typeId: ClarityType, buffer: Buffer): Buffer => {
+  const id = Buffer.from(typeId, 'hex');
+  return Buffer.concat([id, buffer]);
+};
+
+abstract class ClarityValue {
+  abstract type: ClarityType;
+  abstract serialize(): Buffer;
+
+  static deserialize(bufferReader: BufferReader): ClarityValue {
+    return readCV(bufferReader);
+  }
+}
+
+function readCV(bufferReader: BufferReader): ClarityValue {
+    //TODO: make sure you are decoding this properly
+    const type = bufferReader.read(1).toString('hex') as ClarityType;
+
+    switch (type) {
+      case ClarityType.Int:
+        return new IntCV(bufferReader.read(16));
+
+      case ClarityType.UInt:
+        return new UIntCV(bufferReader.read(16));
+
+      case ClarityType.Buffer:
+        const bufferLength = bufferReader.read(4).readUInt32BE(0);
+        return new BufferCV(bufferReader.read(bufferLength));
+
+      case ClarityType.BoolTrue:
+        return new TrueCV();
+
+      case ClarityType.BoolFalse:
+        return new FalseCV();
+
+      case ClarityType.PrincipalStandard:
+        return StandardPrincipalCV.deserialize(bufferReader);
+
+      case ClarityType.PrincipalContract:
+        return ContractPrincipalCV.deserialize(bufferReader);
+
+      case ClarityType.ResponseOk:
+        return new ResponseOkCV(readCV(bufferReader));
+
+      case ClarityType.ResponseErr:
+        return new ResponseErrorCV(readCV(bufferReader));
+
+      case ClarityType.OptionalNone:
+        return new NoneCV();
+
+      case ClarityType.OptionalSome:
+        return new SomeCV(readCV(bufferReader));
+
+      case ClarityType.List:
+        const listLength = bufferReader.read(4).readUInt32BE(0);
+        //TODO: how do I type this?
+        const listContents: ClarityValue[] = [];
+        for (let i = 0; i < listLength; i++) {
+          listContents.push(readCV(bufferReader));
+        }
+        return new ListCV(listContents);
+
+      case ClarityType.Tuple:
+        const tupleLength = bufferReader.read(4).readUInt32BE(0);
+        const tupleContents: { [key: string]: ClarityValue } = {};
```

Comment:
Will this preserve the ordering? Are Clarity tuples ordered?

Reply:
Good question, I don't think this would provide any sort of ordering guarantee. Not sure if Clarity tuples are ordered... @kantai?

Reply:
They aren't ordered types, but they are serialized by blockstack-core in lexicographic order.